### PR TITLE
Adds modular pipe ends

### PR DESCRIPTION
**Pull request by Laboredih123**
[ATMOSPHERICS][INTERNAL][CLEAN]

## About the PR 
Adds modular pipe ends, pipe underlays that can be easily changed to shown, hidden, or exposed. They can also have colors based on what they are connected to.
They use the format SET_PIPE_UNDERLAY(the node being connected to, the dir, size of pipe when exposed which can be "short" "medium" or "long", color of pipe, hide or show the pipe)
Gives it a icon cache so memory and perf is saved.
Adds icon map versions for atmos devices since we kinda don't want a valve with no pipe being shown in SDMM.
Adds subtypes so people don't have to varedit to the map icon they need. Adds these types to maps.
Changes the indicator on the pump and the dual vent so they don't get masked by the pipe color.
Manifolds now have consistent node directions so I don't have to deal with bits on this.

## Why's this needed? 
Makes it so much easier to edit icons, as you don't need to add pipes onto each state.
Improves consistency, as the pipes are centralised in one file.
Helps out mappers a bit so they need to do less varediting.

image

## Changelog 

```changelog
(u)cringe
(+)Atmospheric stuff now has their pipe ends colored, which is pretty rad.
```

### CODE REVIEW

**Comment by Laboredih123** [comment]
hello yes please review well i probably missed something

File: _std/macros/atmos.dm [hunk 1]
```diff
@@ -334,3 +334,24 @@ proc/gas_text_color(gas_id)
 
 #define _LIST_CONCENTRATION_REPORT(GAS, _, NAME, MIXTURE, LIST) LIST += "[NAME]: [round(MIXTURE.GAS / total_moles * 100)]%";
 #define LIST_CONCENTRATION_REPORT(MIXTURE, LIST) APPLY_TO_GASES(_LIST_CONCENTRATION_REPORT, MIXTURE, LIST)
+
+//Possible states are exposed and intact. sizes are short, medium and long.
```
Comment (TobleroneSwordfish) [inline _std/macros/atmos.dm:338]: Are these strings, defines, numbers?

File: _std/macros/atmos.dm [hunk 2]
```diff
@@ -334,3 +334,24 @@ proc/gas_text_color(gas_id)
...
+#define SET_PIPE_UNDERLAY(NODE, DIR, SIZE, COLOUR, HIDDEN) do { \
+	if (UNLINT(HIDDEN)) { \
```
Comment (TobleroneSwordfish) [inline _std/macros/atmos.dm:340]: What's this unlint doing?

File: _std/macros/atmos.dm [hunk 3]
```diff
@@ -334,3 +334,24 @@ proc/gas_text_color(gas_id)
...
+		src.UpdateOverlays(null, "[DIR]"); \
+		break; \
+		}  \
+	var/pipe_state = NODE ? "intact" : "exposed"; \
+	var/pipe_cached = pipe_underlay_cache["[pipe_state]_[DIR]_[SIZE]"]; \
+	if (!pipe_cached) { \
+		pipe_underlay_cache["[pipe_state]_[DIR]_[SIZE]"] = icon('icons/obj/atmospherics/pipes/pipe_underlays.dmi', "[pipe_state]_[NODE ? null : SIZE]", DIR); \
+		pipe_cached = pipe_underlay_cache["[pipe_state]_[DIR]_[SIZE]"]; \
+		} \
+	var/image/pipe_image = mutable_appearance(pipe_cached); \
```
Comment (TobleroneSwordfish) [inline _std/macros/atmos.dm:350]: Is caching and generating a mutable appearance from the cache every time actually any better than just making a new image every time?

File: code/modules/atmospherics/machinery/binary/dp_vent_pump.dm [hunk 4]
```diff
@@ -39,13 +43,18 @@
 	src.hide(T.intact)
 
 /obj/machinery/atmospherics/binary/dp_vent_pump/hide(var/intact) //to make the little pipe section invisible, the icon changes.
-	if(on)
+	var/hide_pipe = intact && issimulatedturf(loc) && level == UNDERFLOOR
+	if(on&&node1&&node2)
```
Comment (TobleroneSwordfish) [inline code/modules/atmospherics/machinery/binary/dp_vent_pump.dm:47]:
```suggestion
	if(on && node1 && node2)
```

File: code/modules/atmospherics/machinery/trinary/filter.dm [hunk 5]
```diff
@@ -13,13 +17,15 @@
 	var/filter_type = "toxins"
 
 /obj/machinery/atmospherics/trinary/filter/update_icon()
-	if(src.node1&&src.node2&&src.node3)
-		src.icon_state = "intact_[src.on?("on"):("off")]"
-	else
-		src.icon_state = "exposed_off"
-
+	if(!(src.node1&&src.node2&&src.node3))
```
Comment (TobleroneSwordfish) [inline code/modules/atmospherics/machinery/trinary/filter.dm:20]:
```suggestion
	if(!(src.node1 && src.node2 && src.node3))
```

File: code/modules/atmospherics/machinery/trinary/mixer.dm [hunk 6]
```diff
@@ -33,17 +37,14 @@
 		MAKE_DEFAULT_RADIO_PACKET_COMPONENT(null, src.frequency)
 
 /obj/machinery/atmospherics/trinary/mixer/update_icon()
-	if(src.node1&&src.node2&&src.node3)
-		icon_state = "intact[src.flipped?"_flipped":""]_[src.on?"on":"off"]"
-	else
-		var/node1_direction = get_dir(src, src.node1)
-		var/node2_direction = get_dir(src, src.node2)
+	if(!(src.node1&&src.node2&&src.node3))
```
Comment (TobleroneSwordfish) [inline code/modules/atmospherics/machinery/trinary/mixer.dm:40]:
```suggestion
	if(!(src.node1 && src.node2 && src.node3))
```

File: code/modules/atmospherics/machinery/trinary/retrofilter.dm [hunk 7]
```diff
@@ -31,12 +31,14 @@
 	var/emagged = FALSE
 
 /obj/machinery/atmospherics/trinary/retrofilter/update_icon()
-	if(src.node1&&src.node2&&src.node3)
-		src.icon_state = "intact_[(src.status & NOPOWER)?("off"):("on")]"
-	else
-		src.icon_state = "" //bad but am lazy to make icons rn planning for a later retrofilter pr
+	if(!(src.node1&&src.node2&&src.node3))
```
Comment (TobleroneSwordfish) [inline code/modules/atmospherics/machinery/trinary/retrofilter.dm:34]:
```suggestion
	if(!(src.node1 && src.node2 && src.node3))
```

File: code/modules/atmospherics/machinery/unary/outlet_injector.dm [hunk 8]
```diff
@@ -104,14 +108,27 @@
...
 /obj/machinery/atmospherics/unary/outlet_injector/hide(var/intact) //to make the little pipe section invisible, the icon changes.
-	if(node)
-		if(on)
-			icon_state = "[intact && issimulatedturf(src.loc) && level == UNDERFLOOR ? "h" : "" ]on"
-		else
-			icon_state = "[intact && issimulatedturf(src.loc) && level == UNDERFLOOR ? "h" : "" ]off"
-	else
-		icon_state = "[intact && issimulatedturf(src.loc) && level == UNDERFLOOR ? "h" : "" ]exposed"
-		on = FALSE
+	var/hide_pipe = intact && issimulatedturf(src.loc) && level == UNDERFLOOR
```
Comment (TobleroneSwordfish) [inline code/modules/atmospherics/machinery/unary/outlet_injector.dm:111]: This feels like it should be a macro maybe?

**Review by TobleroneSwordfish — COMMENTED** [review]
Brief look over the code, thank you for providing an update paths script for secret <3

File: _std/macros/atmos.dm (hunk 1 above)
Comment (Laboredih123) [reply _std/macros/atmos.dm:338]: strings

File: _std/macros/atmos.dm (hunk 2 above)
Comment (Laboredih123) [reply _std/macros/atmos.dm:340]: some devices never have their pipe change visibility, so we set them as FALSE. Unlint is there so checker doesn't complain about it being a constant in an if.

File: _std/macros/atmos.dm (hunk 3 above)
Comment (Laboredih123) [reply _std/macros/atmos.dm:350]: best case, yes. did this with each proc doing it 10000 times.
image

**Review by amylizzle — APPROVED** [review]
Looks way better. Can't see anything broken with it in testing, code seems fine.

**Comment by vortex1942** [comment]
Was the revert of #17334 intentional?

**Comment by Laboredih123** [comment]
> Was the revert of #17334 intentional?

no not particularly

**Comment by Laboredih123** [comment]
mapmerge was being funky so im blaming that

**Pull request merged**